feat: Implement system method `inspect` to satisfy popular demand (#3210)
dfinity/motoko: doc/modules/language-guide/pages/message-inspection.adoc

= Message Inspection
:proglang: Motoko
:company-id: DFINITY

On the {IC}, a canister can selectively

> https://smartcontracts.org/docs/current/references/ic-interface-spec/#ingress-message-inspection[inspect and decline] ingress messages submitted through the https interface:

Author: ```suggestion
_inspect_ and _accept_ or _decline_ ingress messages submitted through the HTTP interface:
```

> A canister can inspect ingress messages before executing them. When the IC receives an update call from a user, the IC will use the canister method `canister_inspect_message` to determine whether the message shall be accepted.
> If the canister is empty (i.e. does not have a Wasm module), then the ingress message will be rejected.
> If the canister is not empty and does not implement `canister_inspect_message`, then the ingress message will be accepted.
>
> In `canister_inspect_message`, the canister can accept the message by invoking
`ic0.accept_message : () -> ()`.
> This function traps if invoked twice.
> If the canister traps in `canister_inspect_message` or does not call `ic0.accept_message`, then the access is denied.
>

> > NOTE: The `canister_inspect_message` is _not_ invoked for query calls, inter-canister calls or calls to the management canister.

Author: ```suggestion
> NOTE: The `canister_inspect_message` is _not_ invoked for HTTP query calls, inter-canister calls or calls to the management canister. 
> NOTE:  Confusingly, a `shared query` function _can_ be called using a regular HTTP call to obtain a certified response: this is why the variant type also includes `shared query` functions.
```

> > -- IC Interface specification

Author: ```suggestion
> -- https://smartcontracts.org/docs/current/references/ic-interface-spec/#ingress-message-inspection[IC Interface Specification]
```

The function is invoked (by the system) on each ingress message. Similar to a query, any side-effects of an invocation are discarded and transient.
A call that traps due to some fault has the same result as returning `false` (message denial).

Unlike other system functions, that have a fixed argument type, the argument type of `inspect_message` depends on the interface of the enclosing actor.
In particular, the formal argument of `inspect_message` is a record of fields of the following types:

* `caller : Principal`: the principal, possibly anonymous, of the caller of the message;
* `arg : Blob` : the raw, binary content of the message argument;
* `msg : <variant>` where `<variant> == {...; #<id>: () -> T; ...}` is a variant of _decoding_ functions,
   one per shared function of the actor called `<id>`. Each variant contains a function that, when applied, returns the argument of function `<id>` as a value of type `T`.

Using a variant, tagged with `#<id>`, allows the return type, `T`, of the decoding function to vary with the identifier, `<id>`, of the shared function.

> Exploiting {proglang}'s subtyping, the formal argument can omit record fields it does not require, or selectively ignore the arguments of particular shared functions, for example,

Author: ```suggestion
Exploiting subtyping, the formal argument can omit record fields it does not require, or selectively ignore the arguments of particular shared functions, for example,
```

= Message Inspection
:proglang: Motoko
:company-id: DFINITY

On the {IC}, a canister can selectively
_inspect_ and _accept_ or _decline_ ingress messages submitted through the HTTP interface:

> > A canister can inspect ingress messages before executing them. When the IC receives an update call from a user, the IC will use the canister method `canister_inspect_message` to determine whether the message shall be accepted.

Reviewer: `canister_inspect_message` ,- yer mother term?

Reviewer: Ah, this is the Wasm-module level interface? Can Motoko actors add such a method? Probably not, since all callables must have a space char in the name.

Author: Yes this is literally quoting from the spec so talking about canisters not actors.

> If the canister is empty (i.e. does not have a Wasm module), then the ingress message will be rejected.
> If the canister is not empty and does not implement `canister_inspect_message`, then the ingress message will be accepted.
>
> In `canister_inspect_message`, the canister can accept the message by invoking
`ic0.accept_message : () -> ()`.
> This function traps if invoked twice.
> If the canister traps in `canister_inspect_message` or does not call `ic0.accept_message`, then the access is denied.
>
> The `canister_inspect_message` is _not_ invoked for HTTP query calls, inter-canister calls or calls to the management canister.
> -- https://smartcontracts.org/docs/current/references/ic-interface-spec/#ingress-message-inspection[IC Interface Specification]

Message inspection mitigates some denial of service attacks,
designed to drain canisters of cycles by placing unsolicited free calls.

> REMARK: You can think of `inspect_message` as providing the "Collect call from <name>. Do you accept charges?" prologue of an old-fashioned, operator-assisted, collect phone call.

Reviewer: `inspect_message` is Motoko lingo.

Author: ```suggestion
REMARK: You can think of method inspection as providing the "Collect call from <name>. Do you accept charges?" prologue of an old-fashioned, operator-assisted, collect phone call.
```

   per shared function, `<id>`, of the actor.
   The variant's tag identifies the function to be called;
   The variant's argument is a function that, when applied, returns the (decoded) argument of the call as a value of type `T`.

Using a variant, tagged with `#<id>`, allows the return type, `T`, of the decoding function to vary with the argument type (also `T`) of the shared function `<id>`.

The variant's argument is a function so that one can avoid the expense of message decoding (when appropriate).

Exploiting subtyping, the formal argument can omit record fields it does not require, or selectively ignore the arguments of particular shared functions, for example,
in order to simply dispatch on the name of a function without inspecting its actual argument.

NOTE:  Confusingly, a `shared query` function _can_ be called using a regular HTTP call to obtain a certified response: this is why the variant type also includes `shared query` functions.

WARNING: An actor that fails to declare system field `inspect_message` will simply accept all ingress messages.

> 

Author: ```suggestion

WARNING: `inspect_message` should **not** be used for definitive access control and any reliable access control checks should be performed within the `shared` functions guarded by `inspect_message`. This is because `inspect_message` is executed by a single replica, without consensus. and its result could be spoofed by a malicious boundary node.
```

[source.no-repl,motoko]
----
include::../examples/InspectArg.mo[tag=inspect-arg]
----

Declining messages by name only, ignoring message arguments (note the use
of type `Any` as message argument variants):

[source.no-repl,motoko]
----
include::../examples/InspectName.mo[tag=inspect-name]
----

A combination of the previous three, specifying the argument types of some variants while ignoring others at type `Any` and using pattern matching to conflate identical cases.

[source.no-repl,motoko]
----
include::../examples/InspectMixed.mo[tag=inspect-mixed]
----

== Tips on authoring `inspect_message`

Implementing `inspect_message` after the fact, once all shared functions of an actor have already been implemented, can be tedious, since you'll need to declare a correctly typed
variant for each shared function.
A simple trick is to first implement the function _incorrectly_, with a `()` argument, compile the code, then use the compiler's error message
to obtain the required argument type.

For example, in the actor from the previous section, incorrectly declaring:

[source.no-repl,motoko]
----
include::../examples/InspectTrick.mo[tag=inspect-trick]
----

> forces the compiler to cough up the expected type below:

Reviewer: reveal? :-)

Author: ```suggestion
forces the compiler to report the expected type below:
```
